File: 39-combination-sum/combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
    ans := make([][]int, 0)
    comb := make([]int, 0)
    n := len(candidates)
    var backtrace func(int, int, []int)
    backtrace = func(t, start int, comb []int) {
        if t == 0 {
            temp := make([]int, len(comb))
            copy(temp, comb)
            ans = append(ans, temp)
            return
        }

        if t < 0 {
            return
        }

        for i := start; i < n; i++ {
            comb = append(comb, candidates[i])
            backtrace(t - candidates[i], i, comb)
            comb = comb[:len(comb)-1]
        }
    }
    backtrace(target, 0, comb)
    return ans
}

File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> comb;
        backtrace(candidates, target, 0, comb, 0, ans);
        return ans;
    }

    void backtrace(vector<int>& candidates, int target, int i, vector<int>& comb, int total, vector<vector<int>>& ans) {
        if (total == target) {
            ans.push_back(comb);
            return;
        }

        if (total > target || i >= candidates.size()) {
            return;
        }

        comb.push_back(candidates[i]);
        backtrace(candidates, target, i, comb, total + candidates[i], ans);
        comb.pop_back();
        backtrace(candidates, target, i + 1, comb, total, ans);
    }
};